examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2169 @@
+    ctx: &impl HasReceiveContext,
+    _host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ContractResult<[u8; 32]> {
+    // Parse the parameter.
+    let param: PermitMessage = ctx.parameter_cursor().get()?;
+
+    // Check that the message was intended for this contract.
+    ensure_eq!(
+        param.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract.into()
+    );
+
+    // Check message is not expired.
+    ensure!(param.timestamp > ctx.metadata().slot_time(), CustomContractError::Expired.into());
+
+    let message_hash = crypto_primitives.hash_sha2_256(&to_bytes(&param)).0;
+
+    match param.payload {
+        // Transfer the tokens.
+        PermitPayload::Transfer(_) => {
+            // Check that the message was intended for this `entry_point`.
+            ensure_eq!(
+                param.entry_point,
+                OwnedEntrypointName::new_unchecked("contract_transfer".into()),
+                CustomContractError::WrongEntryPoint.into()
+            );
+        }
+        // Update the operator.
+        PermitPayload::UpdateOperator(_) => {
+            // Check that the message was intended for this `entry_point`.
+            ensure_eq!(
+                param.entry_point,
+                OwnedEntrypointName::new_unchecked("contract_update_operator".into()),
+                CustomContractError::WrongEntryPoint.into()
+            );
+        }
+    }
+
+    // Return the message hash.
+    Ok(message_hash)
+}
+
+/// Verify an ed25519 signature and allow the transfer of tokens or update of an
+/// operator.
+///
+/// In case of a `transfer` action:
+/// Logs a `Transfer` event and invokes a receive hook function for the
+/// transfer.
+///
+/// In case of a `updateOperator` action:
+/// Logs an `UpdateOperator` event.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - A different nonce is expected.
+/// - The signature was intended for a different contract.
+/// - The signature was intended for a different `entry_point`.
+/// - The signature is expired.
+/// - No public key was registered for the given account.
+/// - The signature can not be validated.
+/// - Fails to log event.
+/// - In case of a `transfer` action: it fails to be executed, which could be
+///   if:
+///     - The `token_id` does not exist.
+///     - The token is not owned by the `from` address.
+///     - The receive hook function calls rejects.
+
+#[receive(
+    contract = "cis3_nft",
+    name = "permit",
+    parameter = "PermitParam",
+    crypto_primitives,
+    mutable,
+    enable_logger
+)]
+fn contract_permit<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ContractResult<()> {
+    // Parse the parameter.
+    let param: PermitParam = ctx.parameter_cursor().get()?;
```

Comment:
There is a `wrongEntryPoint` error in the smart contract already which is thrown when the specified `entry_point` is accidentally swapped around (e.g. somebody wants to call the `transfer` function but provides the `operatorInputParamter`).

https://github.com/Concordium/concordium-rust-smart-contracts/blob/d00aa22d7f6e17111c5af6f6ee211bc947a8537e/examples/cis3-nft-sponsored-txs/src/lib.rs#L326

The embedded schema in the smart contract will specify which input parameters are supported for the `permit` function, so I hope users will make use of the schema and not come up with other inputParameters/functionEntryPoints that they think could be supported. In general, it will be difficult to figure out if this input Parameter deserialization fails due to a small mistake in the inputParameter provided or if the user completely misunderstood it and thought that additional entryPoints/inputParameters are supported. I suggest keeping it for now but I am open to suggestions if there are improvements we can do done the line.

Reply:
While it is helpful, we should not (IMO) rely on the schema to provide information like this.
We could have an endpoint for endpoints supporting sponsored transactions, but a good error code is still useful.